WordPress/gutenberg
# Experiment: Add full page client-side navigation experiment setting

- Part of #60219.

## What?

This pull request adds an **experiment** in Gutenberg -> Experiments to enable full page client-side navigation in your site using the Interactivity API.

**It enables the user experience [showcased in the State Of The Word](https://youtu.be/c7M4mBVgP3Y?t=1659) and in reflected in [the Movies demo](https://wpmovies.dev/)**.

[![State Of The Word 2023](https://img.youtube.com/vi/c7M4mBVgP3Y/0.jpg)](https://www.youtube.com/watch?v=c7M4mBVgP3Y?t=1659s)

Client-side navigation refers to the process of navigating between different pages without the need for a full page refresh. The Interactivity API takes care of fetching the new content from the server and replacing the relevant parts in the DOM, while maintaining the UI state. This approach provides a SPA experience, which is smoother and opens up new possibilities.

One of the limitations of this approach is that all the scripts must be compatible with the Interactivity API, and that's why it can only be enabled in controlled environments until the API is a bit more mature. Additionally, we need to explore deeper how to handle new scripts and styles.

We can differentiate between two types of client-side navigation:

### Region based client-side navigation (existing router implementation)

This type of navigation was included as a first step and it just replaces specific regions of the DOM. This is safer because just the selected regions must be compatible with the Interactivity API. This is used, for example, to [enhance the pagination of the Query Loop block](https://github.com/WordPress/gutenberg/pull/53812).

### Full page client-side navigation (enabled by this experiment)

This type of navigation does replace the whole DOM and not just specific regions. As mentioned above, there might be incompatibilities with other scripts, so it should only be used in controlled environments. This experiment should help to clarify what is needed and how it could be implemented.

## Why?
The idea is to include it as experimental to start testing it and explore what needs to be done to offer it as a site option without the experimental flag.

## How?
There are a few changes in this pull request:
* It adds a new setting in the experiments panel: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-e6bd94e1847aea4b70a6cc1ac9b2471dd3d882f6cbf1ab51438e23b02ca5194fR130-R141). When that option is enabled, it loads conditionally the relevant PHP code.
* If the experiment is enabled, it adds `data-wp-interactive` to all the root blocks: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR31-R72).
* If the experiment is enabled, it registers a new interactivity config setting that can be read in PHP or in JS: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR6-R7).
* It creates a new `interactivity-router` module named `full-csn` to handle the whole site and not only regions: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-c6f39cecb38b1631b9cfa08afbfefd80805da6e69fc2b8eea4868c911dc601e1R1-R205). It is based on the one there was in the initial experiments: [link](https://github.com/WordPress/block-interactivity-experiments/blob/main-wp-directives-plugin/src/runtime/router.js).
* If the experiment is enabled, it registers the new router and, if not, it loads the old one like it is doing in trunk: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-2061d83c07f2215ab0526ac98b601eb910aabdc7a20ac8487be836f54287b31bR24-R39).
* If the experiment is enabled, it adds some directives to all links to prefetch the new page and navigate on click using the router: [link](https://github.com/WordPress/gutenberg/compare/experiment/add-full-client-side-navigation?expand=1#diff-4c31288ce2b84520351c4319e32cbb71215b562d80132a3f7b5441a89bd8795fR9-R27). This should probably not be done by Gutenberg but each site owner should decide where to use it. It is useful now for testing.
* It adapts the query loop block to the new router and loads the script always when the experiment is enabled.

## Testing Instructions
1. Enable the experiment in Gutenberg -> Experiments.
2. Click on links in your site and check that the page is not refreshed.
